Frameworks/Foundation/NSNotificationCenter.mm

```diff
@@ -14,227 +14,307 @@
+- (NSSet<_NSNotificationReceiver*>*)_receiversForName:(NSString*)name observer:(id)observer object:(id)object {
+    NSSet<_NSNotificationReceiver*>* searchSet = _allReceivers;
+    if (name) {
+        searchSet = [self _receiverSetForNotificationName:name createIfMissing:NO];
+        if (!searchSet) {
+            return nil;
+        }
     }
 
-    NSObject* sender = [notification object];
-    int count = CFArrayGetCount((CFArrayRef)arr);
-    unsigned int numSendTo = 0;
-    id* sendTo = (id*)IwMalloc(count * sizeof(id));
-    for (unsigned int i = 0; i < count; i++) {
-        NSNotificationReceiver* observer = (NSNotificationReceiver*)CFArrayGetValueAtIndex((CFArrayRef)arr, i);
-        if (!observer->valid) {
-            continue;
-        }
-        if (observer->notificationSender != NULL && observer->notificationSender != sender) {
-            continue;
-        }
+    __block id blockObserver = observer;
+    __block NSString* blockName = name;
+    __block id blockObject = object;
+    return [searchSet objectsPassingTest:^BOOL(_NSNotificationReceiver* receiver, BOOL* stop) {
+        return (blockObserver == nil || receiver->observer == blockObserver) && (blockObject == nil || receiver->sender == blockObject) &&
+               (blockName == nil || [receiver->name isEqual:blockName]);
+    }];
+}
 
-        sendTo[numSendTo] = observer;
-        [sendTo[numSendTo++] retain];
+// INVARIANT: Called under lock.
+- (void)_addReceiver:(_NSNotificationReceiver*)receiver {
+    StrongId<NSMutableSet<_NSNotificationReceiver*>> receivers = [self _receiverSetForNotificationName:receiver->name createIfMissing:YES];
+    [receivers addObject:receiver];
+    [_allReceivers addObject:receiver];
+}
+
+// INVARIANT: Called under lock.
+- (void)_removeReceivers:(NSSet<_NSNotificationReceiver*>*)receivers forName:(NSString*)name {
+    if ([receivers count] == 0) {
+        return;
     }
 
-    for (unsigned int i = 0; i < numSendTo; i++) {
-        NSNotificationReceiver* observer = sendTo[i];
+    // Update the global receiver table ...
+    [_allReceivers minusSet:receivers];
 
-        if (!observer->valid) {
-            [observer release];
-            continue;
+    // ... and the name lookup tables.
+    if (name) {
+        // Fast path: remove these observers from the provided name's table.
+        StrongId<NSMutableSet<_NSNotificationReceiver*>> nameReceivers = [self _receiverSetForNotificationName:name createIfMissing:NO];
+        if (!nameReceivers) {
+            return;
         }
 
-        if (!observer->block) {
-            [observer->object performSelector:observer->selector withObject:notification];
-        } else {
-            observer->block(notification);
+        [nameReceivers minusSet:receivers];
+
+        if ([nameReceivers count] == 0) {
+            // This can release nameReceivers; it will be retained for the scope of this method.
+            [_observers removeObjectForKey:name];
         }
```

Comment:
I'd like to keep these separate; if the name-free removals are a performance issue, there're some opportunities we can take that don't apply to the other codepath.